### PR TITLE
[JENKINS-27464] Workflow plugin compatibility

### PR DESCRIPTION
**Pull request by zregvart**
Base Jenkins required version is upgraded to 1.580.1 that version
includes the changes required by the Workflow plugin, i.e. using Job
instead of AbstractBuild. For using with workflow plugin one should use
a more recent Jenkins version: latest or LTS 1.596.x.

This allows for using the general SCM support in Jenkins workflow such as:

    import com.ibm.team.build.internal.hjplugin.RTCScm;
    import com.ibm.team.build.internal.hjplugin.RTCScm.BuildType;
   
    checkout changelog: true, poll: true, scm: [$class: 'RTCScm', avoidUsingToolkit: false, buildTool: '...', buildType: new BuildType(RTCScm.BUILD_DEFINITION_TYPE, '...', ""), credentialsId: '...', overrideGlobal: false, serverURI: '...', timeout: 480]

### CODE REVIEW

**Comment by zregvart** [comment]
This pull request is very similar to pull request #5 (deceptively named so I missed it), one addition in this pull request is that unit tests pass on non US locales due to formatting of numbers. Otherwise I believe both pull requests are equal and merging either of them would go towards supporting RTC with Jenkins Workflow plugin.

**Comment by jglick** [comment]
Is this intended to address [JENKINS-27464](https://issues.jenkins-ci.org/browse/JENKINS-27464)? If so, please include `[JENKINS-27464]` in the title, and link to this PR from the issue.

**Comment by jglick** [comment]
> upgraded to 1.580.1 that version includes the changes required by the Workflow plugin, i.e. using Job instead of AbstractBuild

Actually 1.568 is the bare minimum for that, but using an LTS baseline is generally preferable.

**Comment by jglick** [comment]
> `import com.ibm.team.build.internal.hjplugin.RTCScm;`
> `import com.ibm.team.build.internal.hjplugin.RTCScm.BuildType;`

You should not need to import _any_ plugin classes. Rather it should be possible to say something like

```groovy
buildType: [value: 'buildDefinition', buildDefinition: '…']
```

by normalizing the data binding so that `BuildType` is a `Describable<BuildType>` and `getBuildType` returns it. In other words the current binding design is just the wrong way to do it in Jenkins generally; fixing that makes everything work more cleanly in Workflow and other things (YAML project, etc.).

In fact the switch statement points to a bad code smell, too; what you really want is for `BuildType` to be `abstract`, with two subclasses, one for build workspaces, one for build definitions, and a `readResolve` fixing up old settings. That means `config.jelly` can use a standard `f:hetero-radio` (or `f:dropdownDescriptorSelector`) control to select the desired mode, and the flow script would read something like

```groovy
buildType: [$class: 'Definition', value: '…']
```

**Comment by jglick** [comment]
The CI build is broken (not from this change), due to missing top-level POM. I am not sure how one builds this plugin anyway; I could not build either component locally.

File: com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCTask.java [hunk 1]
```diff
@@ -10,17 +10,18 @@
...
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
+import jenkins.MasterToSlaveFileCallable;
+
 /**
  * Base class for tasks running on slaves. Handles common work like debug logging
  */
-public abstract class RTCTask implements Serializable {
+public abstract class RTCTask<T> extends MasterToSlaveFileCallable<T> implements Serializable {
```
Comment (jglick) [inline com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCTask.java]: BTW `implements Serializable` is redundant.

File: com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCChangeLogParser.java [hunk 2]
```diff
@@ -32,87 +29,86 @@
...
-	public ChangeLogSet<? extends Entry> parse(AbstractBuild build,
-			File changelogFile) throws IOException, SAXException {
-		FileInputStream inputStream = new FileInputStream(changelogFile);
-        CharsetDecoder decoder = Charset.forName("UTF-8").newDecoder().onMalformedInput(CodingErrorAction.REPORT).onUnmappableCharacter(CodingErrorAction.REPORT); //$NON-NLS-1$
-		Reader reader = new InputStreamReader(inputStream, decoder);
-		return parse(build, reader);
+	public ChangeLogSet<? extends Entry> parse(Run build,
+			RepositoryBrowser<?> browser, File changeLogFile)
+			throws IOException, SAXException {
+		FileReader changeLogReader = new FileReader(changeLogFile);
```
Comment (jglick) [inline com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCChangeLogParser.java]: Why the change to system encoding from UTF-8?

File: com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCChangeLogParser.java [hunk 3]
```diff
@@ -32,87 +29,86 @@
...
+	public ChangeLogSet<? extends Entry> parse(Run build,
+			RepositoryBrowser<?> browser, Reader changeLogReader)
+			throws IOException, SAXException {
 
-		try {
-			RTCChangeLogSet result = new RTCChangeLogSet(build);
-			Digester2 digester = getDigester();
-			digester.push(result);
-	
-			digester.addSetProperties("changelog"); //$NON-NLS-1$
```
Comment (jglick) [inline com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCChangeLogParser.java]: BTW I recommend trying to minimize diff lines, even at the cost of leaving indentation less than perfect (can always reindent in a separate PR). This is really hard to follow (need to append `?w=1` to URL).

File: com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCBuildResultSetupTask.java [hunk 4]
```diff
@@ -36,7 +36,7 @@
  * to this task since an exception (of any kind) in that will prevent lifecycle
  * info from being linked to the build.
  */
-public class RTCBuildResultSetupTask extends RTCTask implements FileCallable<BuildResultInfo> {
+public class RTCBuildResultSetupTask extends RTCTask<BuildResultInfo> implements FileCallable<BuildResultInfo> {
```
Comment (jglick) [inline com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCBuildResultSetupTask.java]: The `implements FileCallable` is now redundant, right? (Ditto elsewhere.)

File: com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCScm.java [hunk 5]
```diff
@@ -1103,8 +1112,13 @@ public boolean checkout(AbstractBuild<?, ?> build, Launcher arg1,
...
 
-
-		return true;
+	}
+	
+	@Override
+	public SCMRevisionState calcRevisionsFromBuild(Run<?, ?> build,
+			FilePath workspace, Launcher launcher, TaskListener listener)
+			throws IOException, InterruptedException {
+		return SCMRevisionState.NONE;
```
Comment (jglick) [inline com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCScm.java]: BTW this smells like a bug in the plugin which would make polling not work properly. It is supposed to return some representation of the revision from the previous build, so `checkout` can get that `baseline` and properly determine the changelog for a new build.

File: com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCScm.java [hunk 6]
```diff
@@ -1366,4 +1380,20 @@ public String getBuildDefinition() {
 	public String getType() {
 		return super.getType();
 	}
+
+	// lifted from hudson.plugins.git.GitSCM
+    private static Node workspaceToNode(FilePath workspace) {
```
Comment (jglick) [inline com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCScm.java]: You can use `FilePath.toComputer`.

**Comment by jglick** [comment]
I am not sure what `RTCRunListener` is for, but it will not work with Workflow. Needs to be reconsidered.

File: com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCChangeLogParser.java (hunk 3 above)
Comment (zregvart) [inline com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCChangeLogParser.java]: I'm usually very cognisant of whitespace issues, in this case I removed the try-catch block, and the IDE replaced the tabs with spaces. I should have probably just removed the try-catch...

**Comment by zregvart** [comment]
To build the plugin, build from `teamconcert-plugin/com.ibm.team.build.hjplugin/pom.xml`, the `teamconcert-plugin/com.ibm.team.build.hjplugin-rtc` is (from what I gather) a facade to the Rational Team Concert Client. You do not have to build the `...hjplugin-rtc` as the artifact is present in `teamconcert-plugin/com.ibm.team.build.hjplugin/repo` that is declared as a repository in `...hjplugin`. And `hpi:run` et al. is working from `...hjplugin`.

**Comment by zregvart** [comment]
I've fixed the easy stuff as suggested. Will look into implementing a meaningful SCMRevisionState next.

**Comment by amuniz** [comment]
@zregvart the `teamconcert-plugin/com.ibm.team.build.hjplugin-rtc` is an Eclipse plugin. I think is the Eclipse plugin used as RTC client and it has no relation with the Jenkins plugin, actually I wonder why both artifacts are in the same repo.

**Comment by jglick** [comment]
Still missing: a change of `compareRemoteRevisionWith` to the non-deprecated signature. See [this migration guide](https://github.com/jenkinsci/workflow-plugin/blob/master/scm-step/README.md#supporting-workflow-from-an-scm-plugin) for more.

File: com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCScm.java (hunk 5 above)
Comment (zregvart) [inline com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCScm.java]: @jglick you know, I'm not too sure that an implementation of calcRevisionsFromBuild is needed. You see, in RTC you define a build workspace for your Jenkins CI jobs. These are stored on server and RTC in turn then tracks what changes are not accepted in that build workspace from the stream the workspace is following.
This is somewhat analogous to the way cloned git repository can pull changes from the origin, only it happens on the server. The local copy of the workspace in that case is not all that relevant, as the RTC server knows what changes occurred after the build.
See the comment in calcRevisionsFromBuild on https://github.com/zregvart/teamconcert-plugin/blob/master/com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCScm.java#L955-L956.

**Comment by lvaikunt** [comment]
The workflow plugin compatibility is being tracked through https://jazz.net/jazz/resource/itemName/com.ibm.team.workitem.WorkItem/343552 and is being worked upon.

File: com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCScm.java (hunk 5 above)
Comment (jglick) [inline com.ibm.team.build.hjplugin/src/main/java/com/ibm/team/build/internal/hjplugin/RTCScm.java]: OK, might be worth a comment in this method too.

**Comment by zregvart** [comment]
This pull request seems to be obsoleted by ecc68bc924b305bdfdb9d33654fafba05dbda0b1, no need to waste any more effort on it.
